tests/CallbacksTest.test.nut

```diff
@@ -0,0 +1,160 @@
+// MIT License
+//
+// Copyright 2017 Electric Imp
+//
+// SPDX-License-Identifier: MIT
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO
+// EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+@include "./ConnectionManager.lib.nut"
+
+class CallbacksTest extends ImpTestCase {
+
+  cm = null;
+  static TIMEOUT_1 = 5;
+  static TIMEOUT_2 = 10;
+  static TIMEOUT_3 = 15;
+  static TIMEOUT_4 = 20;
+
+  function setUp() {
+    this.info("running setUp");
+    cm = ConnectionManager();
+  }
+
+  /*
+  * sets onDisconnect callback and disconnects device using CM
+  *
+  */
+  function testOnDisconnectAsync() {
+    cm.onDisconnect(function(expected) {
+      this.cm.connect();
+    }.bindenv(this));
+    this.assertTrue(server.isconnected(), "should be connected!");
+    this.cm.disconnect();
+    this.assertTrue(!server.isconnected(), "should NOT be connected!");
+    return Promise(function (resolve, reject) {
+      imp.wakeup(this.TIMEOUT_2, function () {
+        try {
+          this.info("should be online now");
+          this.assertTrue(server.isconnected(), "should be connected again!");
+          resolve();
+        } catch (e) {
+          this.cm.connect(); 
+          reject(e);
+        }
+      }.bindenv(this));
+    }.bindenv(this));
+  }
+
+ /*
+  * sets onConnect callback and disconnects device using CM
+  *
+  */
+  function testOnConnectAsync() {
+    local flag = 0;
+    cm.onConnect(function() {
+      flag++;
+    }.bindenv(this));
+    this.assertTrue((flag == 0), "should NOT be true yet");
+    this.assertTrue(server.isconnected(), "should be connected!");
+    this.cm.disconnect();
```

Comment:
Reusing the same CM object or not "cleaning" it between the tests (not sure which is worse) might not be the best ideas. For example the previous test set the handler:
```
cm.onDisconnect(function(expected) {
   this.cm.connect();
}.bindenv(this));
```
which can interfere with logics here. It may try to reconnect immediately after disconnect.